arrows/mvg/applets/init_cameras_landmarks.cxx

```diff
@@ -0,0 +1,513 @@
+      {
+        if (all_cams.find(id) == all_cams.end())
+        {
+          all_cams[id] = kwiver::vital::camera_sptr();
+        }
+      }
+      camera_map_ptr = std::make_shared<kwiver::vital::simple_camera_map>(all_cams);
+    }
+
+    // If landmark_map_ptr is Null the initialize algorithm will create all
+    // landmarks.  If not Null it will only create landmarks if they are in the
+    // map but Null.  So we need to add placeholders for missing landmarks to
+    // the map.
+    if (landmark_map_ptr)
+    {
+      using kwiver::vital::track_id_t;
+      using kwiver::vital::landmark_map;
+      std::set<track_id_t> track_ids = feature_track_set_ptr->all_track_ids();
+      landmark_map::map_landmark_t all_lms = landmark_map_ptr->landmarks();
+
+      for (auto const& id : track_ids)
+      {
+        if (all_lms.find(id) == all_lms.end())
+        {
+          all_lms[id] = kwiver::vital::landmark_sptr();
+        }
+      }
+      landmark_map_ptr =
+        std::make_shared<kwiver::vital::simple_landmark_map>(all_lms);
+    }
+
+    algorithm->initialize(camera_map_ptr, landmark_map_ptr,
+                          feature_track_set_ptr, sfm_constraint_ptr);
+  }
+};
+
+// ----------------------------------------------------------------------------
+int
+init_cameras_landmarks::
+run()
+{
+  try
+  {
+    d->register_algorithms();
+
+    switch(d->process_command_line(command_args()))
+    {
+      case priv::HELP:
+        std::cout << m_cmd_options->help();
+        return EXIT_SUCCESS;
+      case priv::WRITE:
+        return EXIT_SUCCESS;
+      case priv::FAIL:
+        return EXIT_FAILURE;
+      case priv::SUCCESS:
+        ;
+    }
+
+    if ( d->config == nullptr )
+    {
+      return EXIT_FAILURE;
+    }
+
+    if(d->algorithm == nullptr)
+    {
+      d->initialize();
+    }
+
+    if(d->feature_track_set_ptr == nullptr)
+    {
+      d->load_tracks();
+      if(d->feature_track_set_ptr == nullptr)
+      {
+        LOG_ERROR(main_logger, "There are no feature tracks.");
+        return EXIT_FAILURE;
+      }
+    }
+
+    if(d->sfm_constraint_ptr == nullptr)
+    {
+      d->load_sfm_constraint();
+    }
+
+    d->run_algorithm();
+
+    if(!d->write_cameras())
+    {
+      return EXIT_FAILURE;
+    }
+
+    if(!d->write_landmarks())
+    {
+      return EXIT_FAILURE;
+    }
+
+    return EXIT_SUCCESS;
+  }
+  catch (std::exception const& e)
+  {
+    LOG_ERROR(main_logger, "Exception caught: " << e.what());
+
+    return EXIT_FAILURE;
+  }
+  catch (...)
+  {
+    LOG_ERROR(main_logger, "Unknown exception caught");
+
+    return EXIT_FAILURE;
+  }
+} // run
+
+// ----------------------------------------------------------------------------
+void
+init_cameras_landmarks::
+add_command_options()
+{
+  m_cmd_options->custom_help( wrap_text( "[options]\n" ) );
+
+  m_cmd_options->add_options()
+  ( "h,help",     "Display applet usage" )
+  ( "c,config",   "Configuration file for tool", cxxopts::value<std::string>() )
```

Comment:
I think it would be useful to add command line option here to provide input and output paths.  If specified, these would override what is in the config file.  In fact, the "config" option would rarely be needed at all if a default config is loaded automatically and files paths are specified on the command line.

We should probably also provide defaults for the output file names so that all you really need to provide is the input files on the command line